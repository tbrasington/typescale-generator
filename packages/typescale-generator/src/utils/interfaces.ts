
export interface ViewPortProps {
  width: number;
  fontSize: number;
  typeScale: number;
}

export type TypeScaleProps = {[key: string]: GeneratedItemProps}
export interface GeneratedNamedScalesProps {
  typeScale: TypeScaleProps
}


export interface GeneratedItemProps {
  min: ViewPortProps;
  max?: ViewPortProps;
  clamp: string;
}

export interface RangeProps {
  min: number;
  max: number;
}
export interface GenericKeyValueProps {
  [key: string]: string;
}

export const NamedScales = {
  MAJOR_SECOND: "MAJOR_SECOND",
  MINOR_SECOND: "MINOR_SECOND",
  MAJOR_THIRD: "MAJOR_THIRD",
  MINOR_THIRD: "MINOR_THIRD",
  PERFECT_FOURTH: "PERFECT_FOURTH",
  AUGMENTED_FOURTH: "AUGMENTED_FOURTH",
  PERFECT_FIFTH: "PERFECT_FIFTH",
  GOLDEN_RATIO: "GOLDEN_RATIO",
  OCTAVE: "OCTAVE",
  MAJOR_SIXTH: "MAJOR_SIXTH",
  MINOR_SEVENTH: "MINOR_SEVENTH",
  MAJOR_SEVENTH: "MAJOR_SEVENTH",
};


/*
 * Types for scale enums
 */
export type NamedScalesProps = typeof NamedScales[keyof typeof NamedScales];


export const TypographyScaleValues = {
  [NamedScales.MINOR_SECOND]: { name: "Minor second", value: 1.067 },
  [NamedScales.MAJOR_SECOND]: { name: "Major second", value: 1.125 },
  [NamedScales.MINOR_THIRD]: { name: "Minor third", value: 1.2 },
  [NamedScales.MAJOR_THIRD]: { name: "Major third", value: 1.25 },
  [NamedScales.PERFECT_FOURTH]: { name: "Perfect fourth", value: 1.333 },
  [NamedScales.AUGMENTED_FOURTH]: {
    name: "Augmented fourth",
    value: 1.414,
  },
  [NamedScales.PERFECT_FIFTH]: { name: "Perfect fifth", value: 1.5 },
  [NamedScales.GOLDEN_RATIO]: { name: "Golden ratio", value: 1.618 },
  [NamedScales.MAJOR_SIXTH]: { name: "Major sixth", value: 1.667 },
  [NamedScales.MINOR_SEVENTH]: { name: "Minor seventh", value: 1.778 },
  [NamedScales.MAJOR_SEVENTH]: { name: "Major seventh", value: 1.875 },
  [NamedScales.OCTAVE]: { name: "Octave", value: 2 },
};
