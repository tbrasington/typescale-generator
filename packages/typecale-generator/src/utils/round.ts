
export function round(num: number) {
  return Math.round(num * 100) / 100;
}
