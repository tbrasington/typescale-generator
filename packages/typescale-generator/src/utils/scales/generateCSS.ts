import { GeneratedItemProps, GenericKeyValueProps, TypeScaleProps } from "../interfaces";


export function generateCSS({ scales }: { scales: TypeScaleProps; }) {
  const typeSteps: GenericKeyValueProps = {};

  for (const step of Object.keys(scales)) {
    typeSteps[`--${step}`] = `${scales[`${step}`].clamp};`;
  }

  return `:root ${JSON.stringify(typeSteps, null, "  ").replace(/"/g, "").replace(/,/g, "")}`;
}
