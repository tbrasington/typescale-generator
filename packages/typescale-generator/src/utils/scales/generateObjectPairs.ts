import { GenericKeyValueProps, TypeScaleProps } from "../interfaces";

/**
 * @returns An object of the steps with the min and max unit values, paired with the maxwidth.
 * @example ```js
 *
 * ```
 */

export function generateObjectPairs({ scales }: { scales: TypeScaleProps }) {
   
  const remap = Object.keys(scales).map((item) => {
   const isMaxAvailable = (scales[item].max?.width !== undefined) ? true : false;

    return {
      step: item,
      fontSize: {
        [scales[item].min.width]: scales[item].min.fontSize,
        //@ts-ignore
        ...(isMaxAvailable && {[scales[item].max?.width] : scales[item].max?.fontSize})
      },
    };
  });

  return remap;
}
