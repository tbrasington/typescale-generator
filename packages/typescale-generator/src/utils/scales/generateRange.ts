import { RangeProps } from ".";

/** 
* 
* Generates a range of numbers to use as the steps in the typography scale
* @returns [-2, -1, 0, 1, 2, 3, 4, 5]
*/
export function generateRange({min, max} : RangeProps)  {
  return Array.from(new Array(max + 1 + -1 * min), (_x, i) => i + min);
}
