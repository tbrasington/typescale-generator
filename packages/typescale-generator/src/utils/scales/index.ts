export { generateRange } from "./generateRange";
export { stepValues } from "./stepValues";
export { numberToRem } from "./numberToRem";
export { clampBuilder } from "./clampBuilder";
export { buildTypographyScales } from "./buildTypographyScales";
export { generateObjectPairs } from "./generateObjectPairs";
export { generateCSS } from './generateCSS'
export { generateObject } from './generateObject'