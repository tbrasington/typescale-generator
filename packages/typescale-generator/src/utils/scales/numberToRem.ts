export function numberToRem(px: number): string {
  return `${px / 16}rem`;
}