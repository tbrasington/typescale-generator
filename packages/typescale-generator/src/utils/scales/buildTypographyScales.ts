import { ViewPortProps, GeneratedNamedScalesProps, stepValues } from ".";
import { round } from "../round";
import { clampBuilder } from "./clampBuilder";

/**
 * @returns An array of units and clamped values.
 * ```js
 *    {
 *      typeScale = {
 *         "0": {
 *           clamp : 'clamp(1rem, 0.93rem + 0.36vw, 1.25rem)',
 *           min : {
 *              fontSize: 14.05,
 *              width: 320,
 *              scale : 1.25
 *          },
 *          max : {
 *              fontSize: 17.57,
 *              width: 1440,
 *              scale : 1.25
 *          }
 *        }
 *      }
 *    }
 * ```
 * @example 
 * ```js
*   buildTypographyScales({
      min: {
        width: 320,
        fontSize: 16,
        typeScale: TypographyScaleValues[scale].value,
      },
      max: {
        width: 1440,
        fontSize: 20,
        typeScale: TypographyScaleValues[scale].value,
      },
     range: [-2, -1, 0, 1, 2, 3, 4, 5],
  });
  ```
 */
export function buildTypographyScales(opts: {
  range: number[];
  min: ViewPortProps;
  max?: ViewPortProps;
}) {
  const { min, max, range } = opts;

  const system: GeneratedNamedScalesProps = {
    typeScale: [],
  };

  const minFluidTypeStep = range[0];
  const maxFluidTypeStep = range[range.length - 1];
  for (let i = minFluidTypeStep; i <= maxFluidTypeStep; i++) {
   
    system.typeScale[i] = stepValues({
      min: min,
      max: max,
      step: i,
    })
  }

  return {
    ...system,
  };
}