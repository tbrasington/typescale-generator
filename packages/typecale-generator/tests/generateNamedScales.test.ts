import { generateNamedScales, NamedScales } from "../src/utils/scales"

test('Generate typography scales', () => {
  const typeScales = generateNamedScales(NamedScales["MINOR_SECOND"])
  expect(typeScales).toMatchSnapshot()
})