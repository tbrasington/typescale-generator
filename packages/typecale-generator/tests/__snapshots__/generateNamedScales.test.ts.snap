// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate typography scales 1`] = `
Object {
  "typeScale": Array [
    Object {
      "clamp": "clamp(1rem, 0.93rem + 0.36vw, 1.25rem)",
      "max": Object {
        "fontSize": 20,
        "typeScale": 1.067,
        "width": 320,
      },
      "min": Object {
        "fontSize": 16,
        "typeScale": 1.067,
        "width": 320,
      },
    },
    Object {
      "clamp": "clamp(1.07rem, 0.99rem + 0.38vw, 1.33rem)",
      "max": Object {
        "fontSize": 21.34,
        "typeScale": 1.067,
        "width": 320,
      },
      "min": Object {
        "fontSize": 17.07,
        "typeScale": 1.067,
        "width": 320,
      },
    },
    Object {
      "clamp": "clamp(1.14rem, 1.06rem + 0.41vw, 1.42rem)",
      "max": Object {
        "fontSize": 22.77,
        "typeScale": 1.067,
        "width": 320,
      },
      "min": Object {
        "fontSize": 18.22,
        "typeScale": 1.067,
        "width": 320,
      },
    },
    Object {
      "clamp": "clamp(1.22rem, 1.13rem + 0.43vw, 1.52rem)",
      "max": Object {
        "fontSize": 24.3,
        "typeScale": 1.067,
        "width": 320,
      },
      "min": Object {
        "fontSize": 19.44,
        "typeScale": 1.067,
        "width": 320,
      },
    },
    Object {
      "clamp": "clamp(1.3rem, 1.2rem + 0.46vw, 1.62rem)",
      "max": Object {
        "fontSize": 25.92,
        "typeScale": 1.067,
        "width": 320,
      },
      "min": Object {
        "fontSize": 20.74,
        "typeScale": 1.067,
        "width": 320,
      },
    },
    Object {
      "clamp": "clamp(1.38rem, 1.28rem + 0.49vw, 1.73rem)",
      "max": Object {
        "fontSize": 27.66,
        "typeScale": 1.067,
        "width": 320,
      },
      "min": Object {
        "fontSize": 22.13,
        "typeScale": 1.067,
        "width": 320,
      },
    },
  ],
}
`;
